
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="至少咱还能实践">
      
      
        <meta name="author" content="pzweuj">
      
      
        <link rel="canonical" href="https://pzweuj.github.io/worstpractice/C01_DNA-seq/03.sort_and_mark_duplicates/">
      
      
        <link rel="prev" href="../02.mapping/">
      
      
        <link rel="next" href="../04.call_snp/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.17">
    
    
      
        <title>03. 排序以及标记重复序列 - 生物信息实践</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.bcfcd587.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#_1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="生物信息实践" class="md-header__button md-logo" aria-label="生物信息实践" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            生物信息实践
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              03. 排序以及标记重复序列
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/pzweuj/pzweuj.github.io/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="生物信息实践" class="md-nav__button md-logo" aria-label="生物信息实践" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    生物信息实践
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/pzweuj/pzweuj.github.io/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    前言
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    一、 DNA-seq
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            一、 DNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01.prepare_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    01. 参考基因/数据库/原始数据的准备
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../02.mapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    02. 比对
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    03. 排序以及标记重复序列
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    03. 排序以及标记重复序列
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bam" class="md-nav__link">
    <span class="md-ellipsis">
      转为二进制bam格式
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      排序
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      标记重复序列
    </span>
  </a>
  
    <nav class="md-nav" aria-label="标记重复序列">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#umi" class="md-nav__link">
    <span class="md-ellipsis">
      UMI
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      添加或改写头信息
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      创建索引
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      质量值校正
    </span>
  </a>
  
    <nav class="md-nav" aria-label="质量值校正">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#indel" class="md-nav__link">
    <span class="md-ellipsis">
      indel区域重比对
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_7" class="md-nav__link">
    <span class="md-ellipsis">
      重比对区域校正
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../04.call_snp/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    04. 找变异
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../05.call_somatics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    05. 体细胞突变
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../06.call_cnvs/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    06. CNV突变
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../07.call_fusions/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    07. 融合基因检测
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../08.annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    08. 注释
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../09.visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    09. 可视化
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    二、 RNA-seq
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            二、 RNA-seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C02_RNA-seq/01.prepare_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    01. 参考基因/数据库/原始数据的准备
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C02_RNA-seq/02.mapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    02. 比对
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C02_RNA-seq/03.counting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    03. 计算RPKM/FPKM/TPM/counts
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C02_RNA-seq/04.differential_analysis/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    04. 差异分析
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C02_RNA-seq/05.picture/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    05. 画图
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    三、 mNGS
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            三、 mNGS
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C03_mNGS/01.prepare_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    01. 参考基因/数据库/原始数据的准备
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C03_mNGS/02.remove_host/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    02. 去除宿主基因
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C03_mNGS/03.assemble/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    03. 基因组组装
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C03_mNGS/04.annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    04. 宏基因组注释
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../C03_mNGS/05.pictures/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    05. 可视化
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#bam" class="md-nav__link">
    <span class="md-ellipsis">
      转为二进制bam格式
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_2" class="md-nav__link">
    <span class="md-ellipsis">
      排序
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_3" class="md-nav__link">
    <span class="md-ellipsis">
      标记重复序列
    </span>
  </a>
  
    <nav class="md-nav" aria-label="标记重复序列">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#umi" class="md-nav__link">
    <span class="md-ellipsis">
      UMI
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_4" class="md-nav__link">
    <span class="md-ellipsis">
      添加或改写头信息
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_5" class="md-nav__link">
    <span class="md-ellipsis">
      创建索引
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_6" class="md-nav__link">
    <span class="md-ellipsis">
      质量值校正
    </span>
  </a>
  
    <nav class="md-nav" aria-label="质量值校正">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#indel" class="md-nav__link">
    <span class="md-ellipsis">
      indel区域重比对
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#_7" class="md-nav__link">
    <span class="md-ellipsis">
      重比对区域校正
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="_1">排序和标记重复</h1>
<p>排序和标记重复都是为了后面更好的找变异，从gatk best practice来说，还需要一部加入测序信息的步骤。排序和标记重复均可使用samtools或者picard进行。但是，<a href="https://software.broadinstitute.org/gatk/download">gatk4</a>已经内置了<a href="https://broadinstitute.github.io/picard/">picard</a>的全部功能，这里将会使用gatk4进行这些操作。如果是不是走gatk call snp的流程，其实更建议使用samtools。这一步虽然名字叫排序和标记重复，实际做的事情不止这些，只是排序和标记重复是最重要的。</p>
<h2 id="bam">转为二进制bam格式</h2>
<p>在进行下面的操作之前，先把sam文件转为二进制的bam文件，可以节省空间以及提高效率。
使用samtools进行转化。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>samtools<span class="w"> </span>view<span class="w"> </span>-bSh<span class="w"> </span>B17NC.sam<span class="w"> </span>&gt;<span class="w"> </span>B17NC.bam
</code></pre></div></td></tr></table></div></p>
<p>建议将此步更新为使用<a href="https://lomereiter.github.io/sambamba/">sambamba</a>，速度更快。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>sambamba<span class="w"> </span>view<span class="w"> </span>-S<span class="w"> </span>-h<span class="w"> </span>-f<span class="w"> </span>bam<span class="w"> </span>B17NC.sam<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>-o<span class="w"> </span>B17NC.bam
</code></pre></div></td></tr></table></div></p>
<h2 id="_2">排序</h2>
<p>使用gatk4进行排序。按coordinate进行排序，意思就是按照染色体编号进行排序。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>SortSam<span class="w"> </span>-I<span class="w"> </span>B17NC.bam<span class="w"> </span>-O<span class="w"> </span>B17NC.sorted.bam<span class="w"> </span>-SO<span class="w"> </span>coordinate
</code></pre></div></td></tr></table></div></p>
<p>嫌弃gatk4太慢的，也可以用samtools。事实上，如果对bam文件的要求不怎么严谨，那么这一步已经可以算是最后一步了，当前生产出的bam已经可以用于call snp了。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>samtools<span class="w"> </span>sort<span class="w"> </span>B17NC.bam<span class="w"> </span>-@<span class="w"> </span><span class="m">8</span><span class="w"> </span>-o<span class="w"> </span>B17NC.sorted.bam
</code></pre></div></td></tr></table></div></p>
<p>建议此步使用sambamba，速度更快。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>sambamba<span class="w"> </span>sort<span class="w"> </span>B17NC.bam<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>-o<span class="w"> </span>B17NC.sorted.bam
</code></pre></div></td></tr></table></div></p>
<h2 id="_3">标记重复序列</h2>
<p>标记重复序列实际上修改bam文件的<a href="https://broadinstitute.github.io/picard/explain-flags.html">flag值</a>，重复reads的flag会加上1024用于标记。标记重复是为了去除PCR时产生的大量重复，获得较准确的突变丰度。另外，部分标记重复软件会形成新的tag用于标记，可使用picard/gatk等对tag来进行去重。</p>
<p>这里使用gatk4进行。</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>MarkDuplicates<span class="w"> </span>-I<span class="w"> </span>B17NC.sorted.bam<span class="w"> </span>-O<span class="w"> </span>B17NC.mdup.bam<span class="w"> </span>-M<span class="w"> </span>B17NC.dups.txt
</code></pre></div></td></tr></table></div>
<p>此步可以使用sambamba，速度更快，回报格式与picard/gatk等同。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>sambamba<span class="w"> </span>markdup<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>B17NC.sorted.bam<span class="w"> </span>-o<span class="w"> </span>B17NC.mdup.bam
</code></pre></div></td></tr></table></div></p>
<h3 id="umi">UMI</h3>
<p>对游离DNA进行超高深度测序时一般会加入UMI序列，去重步骤与不加入UMI略有不同。可使用fastp 加上<a href="https://github.com/OpenGene/gencore">gencore</a>的流程进行去重。但是gencore的去重方式是直接去掉而不是标记，并且也不是改写flags值（加上1024）的方式，因此这里使用<a href="http://fulcrumgenomics.github.io/fgbio/">fgbio</a> 加上 gatk （gatk已包含picard的功能）的方式。</p>
<p>IDT有一篇主要<a href="https://sfvideo.blob.core.windows.net/sitefinity/docs/default-source/user-guide-manual/analysis-guideline-variant-calling-data-with-umis.pdf">使用fgbio进行去重的流程</a>，可以参考。而以下的流程仅使用fgbio进行UMI提取，再使用gatk MarkDuplicates进行去重。</p>
<p>先将<strong>过滤后</strong>的fastq转为ubam。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>FastqToSam<span class="w"> </span>-F1<span class="w"> </span>clean.1.fq.gz<span class="w"> </span>-F2<span class="w"> </span>clean.2.fq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-O<span class="w"> </span>sample.ubam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-SM<span class="w"> </span>sample<span class="w"> </span>-PL<span class="w"> </span>illumina<span class="w"> </span>-PU<span class="w"> </span>unit
</code></pre></div></td></tr></table></div></p>
<p>然后提取UMI，以双端3bp在Index的3‘的UMI 为例，用的结构式是3M1S+T，即3bp UMI然后跳1bp，剩余的都是Template。更多的Reads结构写法指导参考<a href="https://github.com/fulcrumgenomics/fgbio/wiki/Read-Structures">此篇</a>。</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code>java<span class="w"> </span>jar<span class="w"> </span>fgbio.jar<span class="w"> </span>ExtractUmisFromBam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-i<span class="w"> </span>sample.ubam<span class="w"> </span>-o<span class="w"> </span>sample.umi.ubam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-r<span class="w"> </span>3M1S+T<span class="w"> </span>3M1S+T<span class="w"> </span>-t<span class="w"> </span>ZA<span class="w"> </span>ZB<span class="w"> </span>-s<span class="w"> </span>RX
</code></pre></div></td></tr></table></div>
<p>然后将ubam比对到参考基因组，要注意的是，由于fgbio默认将Read Group中的ID写为A，所以此时比对最好也是将ID写为A，后续可以用gatk AddOrReplaceReadGroups来修改。主要bwa mem的-p参数必不可少。</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code>samtools<span class="w"> </span>fastq<span class="w"> </span>sample.umi.ubam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>bwa<span class="w"> </span>mem<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>-p<span class="w"> </span>-R<span class="w"> </span><span class="s2">&quot;@RG\tPL:illumina\tPU:unit\tSM:sample\tID:A&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>hg38.fa<span class="w"> </span>/dev/stdin<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>-bSh<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>sample.umi.bam
</code></pre></div></td></tr></table></div>
<p>将ubam和bam合并，使比对后的bam中能带上未比对的ubam中的UMI tag。</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>MergeBamAlignment<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-R<span class="w"> </span>hg38.fa<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-ALIGNED<span class="w"> </span>sample.umi.bam<span class="w"> </span>-UNMAPPED<span class="w"> </span>sample.umi.ubam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-O<span class="w"> </span>sample.umi.merge.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--ALIGNER_PROPER_PAIR_FLAGS<span class="w"> </span><span class="nb">true</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--ATTRIBUTES_TO_RETAIN<span class="w"> </span>XS
</code></pre></div></td></tr></table></div>
<p>最后还是使用MarkDuplicates来去重，记得--BARCODE_TAG RX这个参数。</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>MarkDuplicates<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-I<span class="w"> </span>sample.umi.merge.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-O<span class="w"> </span>sample.marked.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-M<span class="w"> </span>sample.dups.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--CREATE_INDEX<span class="w"> </span><span class="nb">true</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--BARCODE_TAG<span class="w"> </span>RX
</code></pre></div></td></tr></table></div>
<h2 id="_4">添加或改写头信息</h2>
<p>使用gatk4来改变头信息，如果后续用gatk4来进行体细胞突变分析，头信息还是很重要的。当然，如果使用bwa比对时已经用来-R来加入信息，就不需要这一步了。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>AddOrReplaceReadGroups<span class="w"> </span>-I<span class="w"> </span>B17NC.mdup.bam<span class="w"> </span>-O<span class="w"> </span>B17NC.addhead.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-LB<span class="w"> </span>lib1<span class="w"> </span>-PL<span class="w"> </span>illumina<span class="w"> </span>-PU<span class="w"> </span>unit1<span class="w"> </span>-SM<span class="w"> </span>B17NC
</code></pre></div></td></tr></table></div></p>
<h2 id="_5">创建索引</h2>
<p>可以使用gatk4创建索引，也可以使用samtools。
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>BuildBamIndex<span class="w"> </span>-I<span class="w"> </span>B17NC.addhead.bam
</code></pre></div></td></tr></table></div></p>
<p>更建议使用samtools
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>samtools<span class="w"> </span>index<span class="w"> </span>B17NC.addhead.bam
</code></pre></div></td></tr></table></div></p>
<h2 id="_6">质量值校正</h2>
<p>接下来就是使用gatk的质量值校正部分了。gatk4使用机器学习的方式，以各种现有的数据集作为训练集来校正质量值。实际上，在bwa比对后再经过samtools排序的bam文件就已经可以进行call snp了。只不过gatk多一手校正。</p>
<h3 id="indel">indel区域重比对</h3>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>BaseRecalibrator<span class="w"> </span>-I<span class="w"> </span>B17NC.addhead.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-R<span class="w"> </span>Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--known-sites<span class="w"> </span>Mills_and_1000G_gold_standard.indels.b38.vcf.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--known-sites<span class="w"> </span>dbsnp_138.vcf<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--known-sites<span class="w"> </span>1000g_phase.indels.vcf<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-O<span class="w"> </span>B17NC.recal.table
</code></pre></div></td></tr></table></div>
<h3 id="_7">重比对区域校正</h3>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code>gatk<span class="w"> </span>ApplyBQSR<span class="w"> </span>-R<span class="w"> </span>Homo_sapiens_assembly38.fasta<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-I<span class="w"> </span>B17NC.addhead.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--bqsr-recal-file<span class="w"> </span>B17NC.recal.table<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-O<span class="w"> </span>B17NC.final.bam
</code></pre></div></td></tr></table></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2024 pzweuj
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.1e8ae164.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>